subsys/bluetooth/shell/bt.c

```diff
@@ -453,19 +455,34 @@ static int cmd_hci_cmd(const struct shell *shell, size_t argc, char *argv[])
 {
 	u8_t ogf;
 	u16_t ocf;
+	u8_t hex_data[HCI_CMD_MAX_PARAM];
 	struct net_buf *buf = NULL, *rsp;
 	int err;
+	int hex_data_len;
 
+	hex_data_len = 0;
 	ogf = strtoul(argv[1], NULL, 16);
 	ocf = strtoul(argv[2], NULL, 16);
 
 	if (argc > 3) {
+		size_t len;
 		size_t i;
 
-		buf = bt_hci_cmd_create(BT_OP(ogf, ocf), argc - 3);
+		if (strlen(argv[3]) > 2*HCI_CMD_MAX_PARAM) {
+			shell_error(shell, "Data field too large\n");
+			return -ENOEXEC;
+		}
+
+		len = hex2bin(argv[3], strlen(argv[3]), &hex_data[hex_data_len],
+			      sizeof(hex_data) - hex_data_len);
+		if (!len) {
+			shell_error(shell, "HCI command illegal data field\n");
+			return -ENOEXEC;
+		}
 
-		for (i = 3; i < argc; i++) {
-			net_buf_add_u8(buf, strtoul(argv[i], NULL, 16));
+		buf = bt_hci_cmd_create(BT_OP(ogf, ocf), len);
+		for (i = 0; i < len; i++) {
+			net_buf_add_u8(buf, hex_data[i]);
```

Comment:
I think net_buf_add_mem could be used here.

Reply:
Replaced with net_buf_add_mem